src/hotspot/share/gc/shenandoah/heuristics/shenandoahAdaptiveHeuristics.cpp

```diff
@@ -348,137 +426,146 @@ bool ShenandoahAdaptiveHeuristics::should_start_gc() {
+                         byte_size_in_proper_unit(allocation_headroom), proper_unit_for_byte_size(allocation_headroom));
 
-  log_info(gc)("Increased size of %s generation, re-evaluate trigger criteria", _generation->name());
-  return should_start_gc();
+      _last_trigger = RATE;
+      return true;
+    }
+
+    bool is_spiking = _allocation_rate.is_spiking(rate, _spike_threshold_sd);
+    if (is_spiking && avg_cycle_time > allocation_headroom / rate) {
+      log_info(gc)("Trigger (%s): Average GC time (%.2f ms) is above the time for instantaneous allocation rate (%.0f %sB/s)"
+                   " to deplete free headroom (" SIZE_FORMAT "%s) (spike threshold = %.2f)",
+                   _generation->name(), avg_cycle_time * 1000,
+                   byte_size_in_proper_unit(rate), proper_unit_for_byte_size(rate),
+                   byte_size_in_proper_unit(allocation_headroom), proper_unit_for_byte_size(allocation_headroom),
+                   _spike_threshold_sd);
+      _last_trigger = SPIKE;
+      return true;
+    }
+
+    ShenandoahHeap* heap = ShenandoahHeap::heap();
+    if (heap->mode()->is_generational()) {
+      // Get through promotions and mixed evacuations as quickly as possible.  These cycles sometimes require significantly
+      // more time than traditional young-generation cycles so start them up as soon as possible.  This is a "mitigation"
+      // for the reality that old-gen and young-gen activities are not truly "concurrent".  If there is old-gen work to
+      // be done, we start up the young-gen GC threads so they can do some of this old-gen work.  As implemented, promotion
+      // gets priority over old-gen marking.
+
+      size_t promo_potential = heap->get_promotion_potential();
+      size_t promo_in_place_potential = heap->get_promotion_in_place_potential();
+      ShenandoahOldHeuristics* old_heuristics = (ShenandoahOldHeuristics*) heap->old_generation()->heuristics();
+      size_t mixed_candidates = old_heuristics->unprocessed_old_collection_candidates();
+      if (promo_potential > 0) {
+        // Detect unsigned arithmetic underflow
+        assert(promo_potential < heap->capacity(), "Sanity");
+        log_info(gc)("Trigger (%s): expedite promotion of " SIZE_FORMAT "%s",
+                     _generation->name(), byte_size_in_proper_unit(promo_potential), proper_unit_for_byte_size(promo_potential));
+        return true;
+      } else if (promo_in_place_potential > 0) {
+        // Detect unsigned arithmetic underflow
+        assert(promo_in_place_potential < heap->capacity(), "Sanity");
+        log_info(gc)("Trigger (%s): expedite promotion in place of " SIZE_FORMAT "%s", _generation->name(),
+                     byte_size_in_proper_unit(promo_in_place_potential),
+                     proper_unit_for_byte_size(promo_in_place_potential));
+        return true;
+      } else if (mixed_candidates > 0) {
```

Comment:
Yes.  I think it's what we want.  It's basically our "compensation" for the fact that we do not have a truly concurrent old-gen collector.  If the old-collector threads were concurrent, they would be consuming all available idle cycles to take care of the mixed evacuations.

Certainly, this is an area that we can experiment with and auto-tune if appropriate.  My experiments with Phased-Updates workload suggested that if we have a bunch of mixed evacuation work to be done, we want to get it done as possible.  Otherwise, we can get caught flat footed if mutator demand for new allocations spikes.